File: lib/features/auth/models/user_model.dart

class User {
  final String id;
  final String firstName;
  final String lastName;
  final String email;
  final String phone;
  final String role;
  final bool isVerified;
  final Wallet? wallet;
  final String? profilePicture;
  
  User({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.phone,
    required this.role,
    required this.isVerified,
    this.wallet,
    this.profilePicture,
  });
  
  String get fullName => '$firstName $lastName';
  
  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['id'],
      firstName: json['firstName'],
      lastName: json['lastName'],
      email: json['email'],
      phone: json['phone'],
      role: json['role'],
      isVerified: json['isVerified'] ?? false,
      wallet: json['wallet'] != null ? Wallet.fromJson(json['wallet']) : null,
      profilePicture: json['profilePicture'],
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'phone': phone,
      'role': role,
      'isVerified': isVerified,
      'wallet': wallet?.toJson(),
      'profilePicture': profilePicture,
    };
  }
}

class Wallet {
  final double balance;
  final String? virtualAccountNumber;
  final String? virtualAccountName;
  final String? virtualAccountBank;
  
  Wallet({
    required this.balance,
    this.virtualAccountNumber,
    this.virtualAccountName,
    this.virtualAccountBank,
  });
  
  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      balance: (json['balance'] is int) 
          ? (json['balance'] as int).toDouble() 
          : json['balance'],
      virtualAccountNumber: json['virtualAccountNumber'],
      virtualAccountName: json['virtualAccountName'],
      virtualAccountBank: json['virtualAccountBank'],
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'balance': balance,
      'virtualAccountNumber': virtualAccountNumber,
      'virtualAccountName': virtualAccountName,
      'virtualAccountBank': virtualAccountBank,
    };
  }
}

File: Dash_backend/src/models/userModel.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const validator = require('validator');

const userSchema = new mongoose.Schema(
  {
    firstName: {
      type: String,
      required: [true, 'First name is required'],
      trim: true,
    },
    lastName: {
      type: String,
      required: [true, 'Last name is required'],
      trim: true,
    },
    email: {
      type: String,
      required: [true, 'Email is required'],
      unique: true,
      lowercase: true,
      validate: [validator.isEmail, 'Please provide a valid email'],
    },
    phone: {
      type: String,
      required: [true, 'Phone number is required'],
      unique: true,
      validate: {
        validator: function (v) {
          return /^\d{11}$/.test(v); // Nigerian phone format (11 digits)
        },
        message: 'Please provide a valid phone number (11 digits)',
      },
    },
    password: {
      type: String,
      required: [true, 'Password is required'],
      minlength: [8, 'Password must be at least 8 characters'],
      select: false,
    },
    role: {
      type: String,
      enum: ['rider', 'driver', 'admin'],
      default: 'rider',
    },
    profilePicture: {
      type: String,
      default: '',
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    isVerified: {
      type: Boolean,
      default: false,
    },
    createdAt: {
      type: Date,
      default: Date.now,
    },
    updatedAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Virtual for full name
userSchema.virtual('fullName').get(function () {
  return `${this.firstName} ${this.lastName}`;
});

// Virtual populate for wallet
userSchema.virtual('wallet', {
  ref: 'Wallet',
  foreignField: 'user',
  localField: '_id',
  justOne: true,
});

// Pre-save middleware to hash password
userSchema.pre('save', async function (next) {
  if (!this.isModified('password')) return next();
  
  try {
    const salt = await bcrypt.genSalt(12);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (error) {
    next(error);
  }
});

// Method to compare password
userSchema.methods.comparePassword = async function (candidatePassword) {
  return await bcrypt.compare(candidatePassword, this.password);
};

const User = mongoose.model('User', userSchema);

module.exports = User;
